go/database/mpt/proof_test.go

```diff
@@ -0,0 +1,468 @@
+
+	tests := map[string]struct {
+		desc NodeDesc
+	}{
+		"extensionNode wrong path": {&Extension{
+			path: AddressToNibblePath(common.Address{1}, ctxt),
+			next: &Empty{},
+		}},
+		"extensionNode to EmptyNode - path exhausted": {&Extension{
+			path: AddressToNibblePath(address, ctxt),
+			next: &Empty{},
+		}},
+		"extensionNode to EmptyNode": {&Extension{
+			path: AddressToNibblePath(address, ctxt)[0:30],
+			next: &Empty{},
+		}},
+		"different accountNode": {&Extension{
+			path: AddressToNibblePath(address, ctxt)[0:31],
+			next: &Branch{
+				children: Children{
+					AddressToNibblePath(address, ctxt)[31]: &Account{address: common.Address{1}, pathLength: 64, info: AccountInfo{Nonce: common.Nonce{0x01}, Balance: common.Balance{0x02}, CodeHash: common.Hash{0x03}}},
+				}},
+		}},
+		"valueNode key not found": {&Extension{
+			path: AddressToNibblePath(address, ctxt)[0:30],
+			next: &Account{address: address, pathLength: 34, info: AccountInfo{Nonce: common.Nonce{0x01}, Balance: common.Balance{0x02}, CodeHash: common.Hash{0x03}},
+				storage: &Value{key: common.Key{1}, length: 1, value: common.Value{0x01, 0x02, 0x03, 0x04}},
+			},
+		}},
+	}
+
+	extraProof := make(proofDb)
+	extraProof[EmptyNodeEthereumHash] = emptyStringRlpEncoded
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			root, node := ctxt.Build(test.desc)
+			totalProof := createReferenceProof(t, ctxt, &root, node)
+
+			proofWithEmpty := MergeProofs(totalProof, WitnessProof{extraProof})
+			handle := node.GetViewHandle()
+			defer handle.Release()
+
+			rootHash, dirty := handle.Get().GetHash()
+			if dirty {
+				t.Fatalf("expected node to be clean")
+			}
+
+			extractedProof, exists := proofWithEmpty.Extract(rootHash, address, key)
+			if exists {
+				t.Fatalf("proof should not exist")
+			}
+
+			// cannot be proven, but the proof must be still complete
+			if got, want := extractedProof, totalProof; !got.Equals(want) {
+				t.Errorf("unexpected proof: got %v, want %v", got, want)
+			}
+		})
+	}
+}
+
+func TestWitnessProof_Extract_Can_Extract_Terminal_Nodes_In_Proof(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+	address := common.Address{1}
+	nibbles := addressToHashedNibbles(address)
+
+	tests := map[string]struct {
+		trie NodeDesc // < the structure of the trie
+		path []string // < the path to follow to reach the test account
+	}{
+		"empty": {
+			trie: &Tag{"A", &Empty{}},
+			path: []string{},
+		},
+		"wrong account": {
+			trie: &Tag{"A", &Account{}},
+			path: []string{"A"},
+		},
+		"branch without account": {
+			trie: &Tag{"A", &Branch{children: Children{
+				0x1: &Tag{"B", &Empty{}},
+				0x2: &Tag{"C", &Account{address: address}},
+			}}},
+			path: []string{"A"},
+		},
+		"branch with wrong account": {
+			trie: &Tag{"A", &Branch{children: Children{
+				nibbles[0]: &Tag{"B", &Account{address: address, pathLength: 1}},
```

Comment:
Maybe instead of changing the path length change the address?

Reply:
ok, changed